runtime/near-evm-runner/src/interpreter.rs

```diff
@@ -0,0 +1,348 @@
+    // Return the result unmodified
+    let (return_data, apply, gas_left) = match result {
+        Ok(Ok(GasLeft::Known(left))) => (ReturnData::empty(), true, left),
+        Ok(Ok(GasLeft::NeedsReturn { gas_left: left, data, apply_state })) => {
+            (data, apply_state, left)
+        }
+        Ok(Err(err)) => return Err(convert_vm_error(err)),
+        Err(_) => return Err(VMLogicError::EvmError(EvmError::Reverted)),
+    };
+
+    if apply {
+        state.commit_changes(&state_updates.unwrap())?;
+        state.set_code(&address, &return_data.to_vec())?;
+        Ok(ContractCreateResult::Created(address, gas_left))
+    } else {
+        Ok(ContractCreateResult::Reverted(gas_left, return_data))
+    }
+}
+
+pub fn _create<T: EvmState>(
+    state: &mut T,
+    origin: &Address,
+    sender: &Address,
+    value: U256,
+    call_stack_depth: usize,
+    address: &Address,
+    code: &[u8],
+    gas: &U256,
+    evm_gas_config: &EvmCostConfig,
+    chain_id: u128,
+) -> Result<(ExecTrapResult<GasLeft>, Option<StateStore>)> {
+    let mut store = StateStore::default();
+    let mut sub_state = SubState::new(sender, &mut store, state);
+
+    let params = ActionParams {
+        code_address: *address,
+        code_hash: None,
+        code_version: U256::zero(),
+        address: *address,
+        sender: *sender,
+        origin: *origin,
+        gas: *gas,
+        gas_price: 1.into(),
+        value: ActionValue::Transfer(value),
+        code: Some(Arc::new(code.to_vec())),
+        data: None,
+        action_type: ActionType::Create,
+        params_type: vm::ParamsType::Embedded,
+    };
+
+    sub_state.transfer_balance(sender, address, value)?;
+
+    let mut ext = NearExt::new(
+        *address,
+        *origin,
+        &mut sub_state,
+        call_stack_depth + 1,
+        false,
+        evm_gas_config,
+        chain_id,
+    );
+    ext.info.gas_limit = U256::from(gas);
+    ext.schedule = Schedule::new_constantinople();
+
+    let instance = Factory::default().create(params, ext.schedule(), ext.depth());
+
+    // Run the code
+    let result = instance.exec(&mut ext);
+    Ok((result, Some(store)))
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn call<T: EvmState>(
+    state: &mut T,
+    origin: &Address,
+    sender: &Address,
+    value: Option<U256>,
+    call_stack_depth: usize,
+    contract_address: &Address,
+    input: &[u8],
+    should_commit: bool,
+    gas: &U256,
+    evm_gas_config: &EvmCostConfig,
+    chain_id: u128,
+) -> Result<MessageCallResult> {
+    run_and_commit_if_success(
+        state,
+        origin,
+        sender,
+        value,
+        call_stack_depth,
+        ActionType::Call,
+        contract_address,
+        contract_address,
+        input,
+        false,
+        should_commit,
+        gas,
+        evm_gas_config,
+        chain_id,
+    )
+}
+
+pub fn delegate_call<T: EvmState>(
+    state: &mut T,
+    origin: &Address,
```

Comment:
Isn't the signature long enough already to have smth like `ApplyState` used here?

Reply:
Good point. Worth refactoring it out. It was probably code written by Summa and we just moved it here from the Wasm contract. I'll create an follow up issue for this: https://github.com/near/nearcore/issues/3677